### PR TITLE
Add an on-demand Iris vs RCCL CCL benchmark job

### PR DESCRIPTION
**Pull request by nirvedhmeshram**
## Summary

Adds an on-demand job that benchmarks the `iris.ccl` collectives against RCCL and publishes a comparison table and graphs. Manual trigger or post-submit on `main` — deliberately **not** on pull requests.

## This existed before and was lost

`benchmark/ccl/comprehensive_sweep.py` swept the collectives with a `--benchmark_rccl` flag, and the per-op scripts reported `rccl_ms` / `rccl_bandwidth_gbps` / `rccl_ratio_percent`. It was deleted in 15bba614 when the benchmarks migrated to `iris.bench` (#486), and the RCCL paths went with it.

What survived is `benchmark/ccl/plot_sweep_results.py` — still documented as *"subplots comparing Iris vs RCCL bandwidth for each collective operation"*, still reading a CSV that nothing has produced since. This gives it a real input again.

## Approach

Rather than restore the old subprocess driver, each CCL benchmark gets a `backend` axis:

```python
@bench.axis("backend", ["iris", "rccl"])
```

Both series then come out of **one sweep over identical shapes through the same timing harness**, and the framework's existing CSV already carries the comparison. The process group is already `nccl`, so RCCL needs no extra setup. `preamble_fn` runs outside the timed region, so the `copy_` needed to give `dist.all_reduce` out-of-place semantics does not pollute the measurement.

## Two judgement calls worth reviewing

**The RCCL baseline uses ordinary torch tensors, not Iris heap memory.** The symmetric heap is allocated with the flags RMA requires, so timing RCCL against it would measure Iris's allocation choice rather than RCCL. Someone deciding between the two would call RCCL on normal tensors, so that is what is measured.

**`reduce_scatter` is excluded.** `iris.ccl.reduce_scatter` is `(M, N) -> (M, N)` — "each rank reduces its assigned tiles, stores locally" — whereas `torch.distributed.reduce_scatter_tensor` is `(W*M, N) -> (M, N)`. Publishing those side by side would compare two different operations. Happy to add it if the intended correspondence is different from how I read it.

`all_to_all` is included with a documented caveat: Iris splits `(M, N*W)` along dim 1, `dist.all_to_all_single` splits along dim 0, so the RCCL path uses `(M*W, N)`. Identical communication pattern and total bytes — `W-1` peer messages of `M*N` elements — only the layout differs.

## Cost

Not on `pull_request` by design. Four collectives across two backends is enough work to crowd out the test jobs on a shared 8-GPU pool. Ranks, sizes and repeat count are `workflow_dispatch` inputs; the default sweep is deliberately small (`M=N ∈ {1024, 4096, 16384}`, fp16, 8 ranks) and can be widened per run without editing anything.

## Output

The job writes the table into the run summary and uploads CSVs and PNGs as artifacts (30-day retention):

| Operation | Size (MiB) | Iris (GB/s) | RCCL (GB/s) | Iris/RCCL |
|---|---|---|---|---|
| all_reduce | 2.00 | 120.5 | 95.2 | 1.27x |

## Test plan

The plotting and table path is verified against the real `iris.bench` CSV schema, including that `skipped=True` rows are dropped and a missing RCCL point renders as `—` rather than a bogus ratio.

**The sweep itself has not been run on hardware yet** — that needs the 8-GPU runner, so the first `workflow_dispatch` is the real test. Worth knowing before trusting any numbers it produces.

### CODE REVIEW

File: benchmark/ccl/bench_all_reduce.py [hunk 1]
```diff
@@ -5,28 +5,51 @@
 """Benchmark for iris-ccl all-reduce collective."""
 
 import torch
+import torch.distributed as dist
 import iris.bench as bench
 from iris.ccl import Config
 
 
+# The RCCL baseline deliberately uses ordinary torch tensors rather than Iris
```
Comment (drprajap) [inline benchmark/ccl/bench_all_reduce.py:13]: These are repeated in every bench_*.py, doesn't seem necessary, may be find central place in READMe to describe this?

File: benchmark/ccl/plot_sweep_results.py [hunk 2]
```diff
@@ -1,183 +1,182 @@
...
-    with open(csv_file, "r") as f:
-        reader = csv.DictReader(f)
-        for row in reader:
-            operation = row["operation"]
-            comm_sms = int(row["comm_sms"])
-            iris_bw = float(row["iris_bandwidth_gbps"])
 
-            data[operation]["comm_sms"].append(comm_sms)
-            data[operation]["iris_bw"].append(iris_bw)
+# Matches _dtype_str in iris/bench/_runner.py, which writes the short name.
```
Comment (drprajap) [inline benchmark/ccl/plot_sweep_results.py:66]: https://github.com/ROCm/iris/blob/5b46b8ec2f8b60af977467c5dbfeb4838763854b/iris/bench/_runner.py
there is gap in how results are aggregated. seems like Iris’s runner.py publishes rank 0’s mean timing. 

The way we did  it in triton-shmem harness , it gathers timings from every rank and reports the slowest rank’s median plus rank skew. Both backends in this PR use the same Iris timer, but rank 0 can miss a bottleneck on another GPU.

**Review by drprajap — CHANGES_REQUESTED** [review]

File: .github/scripts/run_ccl_benchmark.sh [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+GPU_DEVICES=${GPU_DEVICES:-"0,1,2,3,4,5,6,7"}
+
+echo "[CCL-BENCHMARK] ranks=${NUM_RANKS} sizes=${SIZES} n_repeat=${N_REPEAT}"
+echo "[CCL-BENCHMARK] Using GPUs: $GPU_DEVICES"
+
+mkdir -p ccl_results
+
+# The default axes sweep M and N over 2^10..2^14 independently, which is 25
+# shape combinations per op per backend. Pin M=N to the requested sizes instead:
+# a square sweep along the diagonal, which is what the comparison plot shows.
```
Comment (drprajap) [inline .github/scripts/run_ccl_benchmark.sh:30]: why did we choose perfect square shapes?

File: benchmark/ccl/bench_all_gather.py [hunk 4]
```diff
@@ -5,25 +5,46 @@
...
+# would call RCCL on normal tensors, so that is what is measured.
+def _torch_tensor(ctx, shape, dtype):
+    return torch.zeros(shape, dtype=dtype, device=f"cuda:{ctx.get_rank()}")
+
+
 @bench.register
 @bench.axis("num_ranks", [2, 4, 8])
 @bench.axis("M", bench.power_of_two(10, 14))
 @bench.axis("N", bench.power_of_two(10, 14))
 @bench.axis("dtype", [torch.float16, torch.bfloat16])
```
Comment (drprajap) [inline benchmark/ccl/bench_all_gather.py:26]: can we include fp8 too?

File: benchmark/ccl/bench_all_gather.py [hunk 5]
```diff
@@ -5,25 +5,46 @@
...
+
+    if state["backend"] == "rccl":
+        t_in = _torch_tensor(ctx, (M, N), dtype)
+        t_out = _torch_tensor(ctx, (world_size * M, N), dtype)
+        t_in.fill_(float(ctx.get_rank() + 1))
+        state.exec(
+            lambda: dist.all_gather_into_tensor(t_out, t_in),
+            preamble_fn=lambda: t_out.zero_(),
+        )
+        return
```
Comment (drprajap) [inline benchmark/ccl/bench_all_gather.py:42]: This seems common among multiple benchmark files, may be reorganize in util or somewhere?

File: benchmark/ccl/plot_sweep_results.py (hunk 2 above)
Comment (nirvedhmeshram) [reply benchmark/ccl/plot_sweep_results.py]: fixed

File: benchmark/ccl/bench_all_gather.py (hunk 4 above)
Comment (nirvedhmeshram) [reply benchmark/ccl/bench_all_gather.py]: there is no RCCL fp8 available so this would not have a baseline, happy to add if we just want to see them by themselves though. Let me know.

File: benchmark/ccl/bench_all_gather.py (hunk 5 above)
Comment (nirvedhmeshram) [reply benchmark/ccl/bench_all_gather.py]: done.

File: .github/scripts/run_ccl_benchmark.sh (hunk 3 above)
Comment (nirvedhmeshram) [reply .github/scripts/run_ccl_benchmark.sh]: fixed to a sweep now following what we had in triton-shmem.

**Comment by nirvedhmeshram** [comment]
These are the graphs this CI is expected to generate (only on demand currently)

iris-vs-rccl-full
iris-vs-rccl-full_latency
iris-vs-rccl-full_speedup
